### DISCUSSION TITLE
Eager loading of aliased relationship

### DISCUSSION THREAD

**Discussion by michael-nml** [sqlalchemy/sqlalchemy#9589] [Usage Questions] [answered]
Hi there, I'm looking for some advice in eagerly loading an aliased relationship. Apologies if this has been asked before; I could not immediately find it.

Using a slightly modified version of [the example from the docs](https://docs.sqlalchemy.org/en/20/orm/join_conditions.html#row-limited-relationships-with-window-functions) to illustrate:

``` python
from sqlalchemy import Integer, ForeignKey, and_, func, select
from sqlalchemy.orm import DeclarativeBase, aliased, mapped_column, relationship


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"
    id = mapped_column(Integer, primary_key=True)

    bs = relationship(lambda: B, lazy="joined", back_populates="a")


class B(Base):
    __tablename__ = "b"
    id = mapped_column(Integer, primary_key=True)
    a_id = mapped_column(ForeignKey("a.id"))

    a = relationship(A, lazy="joined", back_populates="bs")


partition = select(
    B, func.row_number().over(order_by=B.id, partition_by=B.a_id).label("index")
).alias()

partitioned_b = aliased(B, partition)

A.partitioned_bs = relationship(
    partitioned_b, primaryjoin=and_(partitioned_b.a_id == A.id, partition.c.index < 10),
    lazy="joined"
)

print("Loading 'A'".center(80, '-'))
print(select(A))
print("")
print("Loading 'B'".center(80, '-'))
print(select(B))
```

Output:
```
----------------------------------Loading 'A'-----------------------------------
SELECT a.id, b_1.id AS id_1, b_1.a_id, anon_1.id AS id_2, anon_1.a_id AS a_id_1
FROM a LEFT OUTER JOIN b AS b_1 ON a.id = b_1.a_id LEFT OUTER JOIN (SELECT b.id AS id, b.a_id AS a_id, row_number() OVER (PARTITION BY b.a_id ORDER BY b.id) AS index
FROM b) AS anon_1 ON anon_1.a_id = a.id AND anon_1.index < :index_1

----------------------------------Loading 'B'-----------------------------------
SELECT b.id, b.a_id, a_1.id AS id_1
FROM b LEFT OUTER JOIN a AS a_1 ON a_1.id = b.a_id
```

For the most part this works as expected:
* Loading `A` populates both the `bs` and `partitioned_bs` attributes using two joins to the child table.
* Loading `B` populates the `a` attribute using a join to the parent table.

However, in my use case I want to use an instance of `B` to access `partitoned_bs` on the parent object, i.e. `b.a.paritioned_bs` where `b` is an instance of `B`. The `partitioned_bs` attribute wasn't eagerly loaded as visible in the `Loading 'B'` output above.

Normally this wouldn't be a significant issue, but as I'm working in async context this attribute access triggers `MissingGreenlet` errors. Additionally the abstractions in place make it difficult to perform an explicit load of the aliased relationship, so I'm hoping to configure the needed behaviour at the mapper level.

I'm guessing the current behaviour is in place to avoid an endless lookup (load B -> load A -> load B -> load A -> ...). Which makes sense in the case of back-populating relationships. Is there some additional configuration I can use to trigger the eager load in the context of an aliased relationship?

**Comment by zzzeek** [comment, marked as answer]
joined eager loading is not a good choice for loading cycles since it creates for very long LEFT OUTER JOIN chains that are not efficient, however you should be able to set [join_depth](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.join_depth) to a number like 2 and have it do some more levels here.

If you break up the strategies here between "selectinload" for the collection (which is better) and "joinedload" for the many-to-one, it should handle the longer chain of loads, since the recursion detection should be a bit more broken up, but you can also add join_depth for selectinload.

**Reply by michael-nml** [reply]
Perfect, that's exactly what I needed. Thank you for the very quick reply and your overall work on sqlalchemy! Loving the typing support in 2.0.

Also, yes, I'll use "selectinload" for the many-to-one. Joined load was easier for the sake of generating the SQL in the example 😅

**Comment by michael-nml** [comment]
@zzzeek I wasn't sure I understood your feedback correctly about "selectinload", as you seemed to suggest breaking up the load strategies would work without setting `join_depth`. So I did some more tests, and found what may be a bug.

Amended example code:
```python
import asyncio
from sqlalchemy import Integer, ForeignKey, and_, func, select
from sqlalchemy.orm import DeclarativeBase, aliased, mapped_column, relationship
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"
    id = mapped_column(Integer, primary_key=True)

    bs = relationship(lambda: B, lazy="selectin", back_populates="a")


class B(Base):
    __tablename__ = "b"
    id = mapped_column(Integer, primary_key=True)
    a_id = mapped_column(ForeignKey("a.id"))

    a = relationship(A, lazy="joined", back_populates="bs")


partition = select(
    B, func.row_number().over(order_by=B.id, partition_by=B.a_id).label("index")
).alias()

partitioned_b = aliased(B, partition)

A.partitioned_bs = relationship(
    partitioned_b, primaryjoin=and_(partitioned_b.a_id == A.id, partition.c.index < 10),
    lazy="selectin",
    viewonly=True
)


async def main(select_parent: bool):
    engine = create_async_engine("postgresql+asyncpg://test:test@localhost:5432/test", echo=True)

    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    async with AsyncSession(engine, expire_on_commit=False) as session:
        async with session.begin():
            a = A()
            a.bs = [B() for _ in range(5)]
            session.add(a)

        async with session.begin():
            if select_parent:
                results = await session.scalars(select(A))
                a = results.first()
                print("Found:", a.partitioned_bs)
            else:
                results = await session.scalars(select(B))
                b = results.first()
                print("Found:", b.a.partitioned_bs)


asyncio.run(main(select_parent=True))
```

Using `sqlalchemy==2.0.8` this results in an infinite loop when using `select_parent=True`. When setting `select_parent=False`, it results in a  `MissingGreenlet` error. Setting `join_depth=2` on the `A.partitioned_bs` relationship makes it works as intended, regardless of the `select_parent` setting.

In my case I can set `join_depth=2` and all is well, but the infinite loop when it is not set seems like a bug.

**Reply by zzzeek** [reply]
when you said "endless loop" I assumed you meant "recursion overflow" but nope, taht's a real endless loop :).  it's in 1.4 also.   #9590 will fix that
